### ISSUE TITLE
Electron build doesn't work properly

### ISSUE THREAD

**Issue by Fry98** [cmake-js/cmake-js#222]
Hi,
not sure if I'm doing something wrong on my end but when I try to build my Electron C++ addon with CMake.js I end up with some weird issues. The biggest one is that the app instatnly crashes when I call `args[0]->IsString()` during function argument validation. There doesn't seem to be a problem with the other validation methods like `IsNumber()` or `IsStringObject()`. After I toggled on the `--enable-logging` flag in Electron I was able to figure out that this is the error Electron crashes on:
`[13508:1016/202149.533:ERROR:crashpad_client_win.cc(799)] not connected`.

Another weird behavior is that when I'm logging something with `std::cout` it only actually gets printed when compiling in Debug mode. Neither of these issues appeared when I tried building with `node-gyp`.

My `CMakeLists.txt` file is just the very basic configuration:
```
cmake_minimum_required (VERSION 3.5)
set(CMAKE_CXX_STANDARD 17)
project(test)

include_directories(${CMAKE_JS_INC})
add_library(${PROJECT_NAME} SHARED main.cpp ${CMAKE_JS_SRC})
set_target_properties(${PROJECT_NAME} PROPERTIES PREFIX "" SUFFIX ".node")
target_link_libraries (${PROJECT_NAME} ${CMAKE_JS_LIB})
```
and this is my setup in `package.json`:
```
"cmake-js": {
    "runtime": "electron",
    "runtimeVersion": "10.1.3",
    "arch": "x64"
}
```

Am I missing something or is this some sort of internal CMake.js issues? Thanks.

**Comment by mastergberry** [comment]
Most likely related to https://github.com/electron/electron/issues/22799

I am going to try and sit down this next week or two and debug this problem to see what the difference is between cmake-js and node-gyp with the compiler flags etc.

**Comment by mastergberry** [comment]
I attempted this morning to find the difference between a `cmake-js` and `node-gyp` module, but I didn't have any success, even after manually overriding most of the options in the `.vcxproj` file.

Maybe @unbornchikken can take a look at why `cmake-js`/`cmake` no longer work with electron 9+

If this can't be resolved soonish we will probably be forced to go back to `node-gyp` as we need other fixes from higher versions of `electron` and `8.5.2` won't be supported forever.

**Comment by Fry98** [comment]
Thanks for the feedback. I actually didn't know this was an Electron v9+ problem only. For now I've downgraded to v8.x but hopefully this gets resolved soon.

**Comment by mastergberry** [comment]
Hey @Fry98 are you using `nan` by any chance? https://github.com/nodejs/nan

**Comment by Fry98** [comment]
No, I'm not. I even specifically tested this bug with the most basic addon possible to make sure it is actually caused by CMakeJS and not any other deps.

**Comment by mastergberry** [comment]
I actually migrated my project today from nan -> node-addon-api and suddenly saw all of my problems disappear.

**Comment by Fry98** [comment]
Huh, interesting. Well... I was just thinking about migrating to node-addon-api so I guess I'll give that a shot and see what happens.

**Comment by Fry98** [comment]
So I finally switched over to node-addon-api and it does seem like everything is working properly now. Thanks for your advice! 👍

**Comment by vadim-termius** [comment]
Since Electron 9 it uses the patch to enable pointer compression in v8 and its node.js: https://github.com/electron/electron/pull/21468/files#diff-068245f8862a4eeb85c645acdff8f9f1ce4dc946225f623c366d35bfcf38d4cf

Now all native node modules should be built with the `V8_COMPRESS_POINTERS` macro defined only on 64-bit platforms and the `V8_31BIT_SMIS_ON_64BIT_ARCH` macro defined on all platforms (which seems counterintuitive, but in the patch this is defined by default)

Those macros don't affect node-addon-abi, that is why it works fine.

cc @unbornchikken @mastergberry

**Comment by mastergberry** [comment]
Thanks for the information @vadim-termius ...just wish this was more well documented and I didn't waste months on this =(

Still doesn't hurt to move over to the new API for native modules that did not exist years ago...since this one seems to be the new future for native modules with nodejs.

**Comment by vadim-termius** [comment]
Also, Electron 11 now requires `V8_REVERSE_JSARGS`: https://github.com/electron/electron/blob/4f281e3d31710a72f8922419a28c82210b581d28/patches/node/fix_add_v8_enable_reverse_jsargs_defines_in_common_gypi.patch

**Comment by mastergberry** [comment]
So both the `V8_REVERSE_JSARGS` and `V8_COMPRESS_POINTERS` defines should be done when compiling native modules?

**Comment by vadim-termius** [comment]
For Electron 9+ `V8_31BIT_SMIS_ON_64BIT_ARCH` should be defined.

For Electron 9+ for `arm64` and for `x64` `V8_COMPRESS_POINTERS` should also be defined.

For Electron 11+ `V8_REVERSE_JSARGS` should also be defined.

**Comment by lilong7676** [comment]
Help, Can anyone help to read this issue  electron/electron#27205 , I don't know if that has anything to do with it

**Comment by smalls12** [comment]
> 
> 
> For Electron 9+ `V8_31BIT_SMIS_ON_64BIT_ARCH` should be defined.
> 
> For Electron 9+ for `arm64` and for `x64` `V8_COMPRESS_POINTERS` should also be defined.
> 
> For Electron 11+ `V8_REVERSE_JSARGS` should also be defined.

After several days of trying to run my native module, built with cmake-js, in electron, this finally was the breakthrough.
Can this be pinned somewhere ?

And thanks! @vadim-termius

**Comment by caiiiycuk** [comment]
I used this in CMakeLists.txt to build for electron 18+ and x64:
```
add_definitions(
    -DV8_COMPRESS_POINTERS
    -DV8_REVERSE_JSARGS
    -DV8_COMPRESS_POINTERS_IN_ISOLATE_CAGE
)
```

**Comment by GitMurf** [comment]
> This is what I use in our native project building node modules

@miniak I am having same problem with node-canvas (https://github.com/Automattic/node-canvas/) which uses NAN instead of Node-API. Do you know if its possible to put together a similar solution for that as you have above? Thanks!
